File: backend/app/routers/email_sync.py
"""
Email sync routes: run sync, list suggestions, act on suggestions.
Supports Gmail and Yahoo Mail.
Uses the modular email_pipeline for filtering, scoring, and classification.
"""

from datetime import datetime, timedelta, timezone

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ..database import get_db
from ..models import ApplicationUpdateSuggestion, GmailToken, Job, YahooToken
from ..schemas import SuggestionAction, SuggestionResponse
from ..services.email_pipeline import process_email
from ..services.gmail_service import fetch_recent_emails as gmail_fetch
from ..services.gmail_service import refresh_access_token as gmail_refresh
from ..services.yahoo_service import fetch_recent_emails as yahoo_fetch
from ..services.yahoo_service import refresh_access_token as yahoo_refresh

router = APIRouter(prefix="/email-sync", tags=["email-sync"])


async def _get_valid_gmail_token(db: AsyncSession) -> GmailToken | None:
    """Return a valid Gmail token, refreshing if needed. Returns None if not connected."""
    result = await db.execute(select(GmailToken).order_by(GmailToken.id.desc()).limit(1))
    token = result.scalars().first()
    if not token:
        return None

    if token.token_expiry and token.token_expiry < datetime.now(timezone.utc) + timedelta(minutes=5):
        try:
            refreshed = await gmail_refresh(token.refresh_token)
            token.access_token = refreshed["access_token"]
            token.token_expiry = datetime.now(timezone.utc) + timedelta(
                seconds=refreshed.get("expires_in", 3600)
            )
            if "refresh_token" in refreshed:
                token.refresh_token = refreshed["refresh_token"]
            await db.commit()
            await db.refresh(token)
        except Exception:
            return None

    return token


async def _get_valid_yahoo_token(db: AsyncSession) -> YahooToken | None:
    """Return a valid Yahoo token, refreshing if needed. Returns None if not connected."""
    result = await db.execute(select(YahooToken).order_by(YahooToken.id.desc()).limit(1))
    token = result.scalars().first()
    if not token:
        return None

    if token.token_expiry and token.token_expiry < datetime.now(timezone.utc) + timedelta(minutes=5):
        try:
            refreshed = await yahoo_refresh(token.refresh_token)
            token.access_token = refreshed["access_token"]
            token.token_expiry = datetime.now(timezone.utc) + timedelta(
                seconds=refreshed.get("expires_in", 3600)
            )
            if "refresh_token" in refreshed:
                token.refresh_token = refreshed["refresh_token"]
            await db.commit()
            await db.refresh(token)
        except Exception:
            return None

    return token


async def _process_emails(
    db: AsyncSession,
    emails: list[dict],
    jobs: list[dict],
    existing_ids: set,
    id_field: str,
    min_confidence: str = "HIGH",
) -> tuple[int, int]:
    """Run emails through the pipeline and create suggestions."""
    new_suggestions = 0
    skipped = 0

    for email_data in emails:
        msg_id = email_data["id"]

        if msg_id in existing_ids:
            skipped += 1
            continue

        # Run through the full pipeline
        result = process_email(email_data, jobs, min_confidence=min_confidence)
        if result is None:
            continue

        kwargs = {
            "application_id": result.matched_application_id,
            "detected_type": result.classification,
            "suggested_status": result.suggested_status,
            "confidence_score": result.confidence,
            "email_subject": result.email_subject,
            "email_sender": result.email_sender,
            "email_date": result.email_date,
            "email_snippet": result.email_snippet,
            "state": "pending",
        }
        kwargs[id_field] = msg_id

        db.add(ApplicationUpdateSuggestion(**kwargs))
        new_suggestions += 1

    return new_suggestions, skipped


@router.post("/run")
async def run_email_sync(
    confidence: str = "HIGH",
    db: AsyncSession = Depends(get_db),
):
    """
    Fetch recent emails from all connected providers,
    run them through the detection pipeline, and create suggestions.

    Query params:
        confidence: minimum confidence level — "HIGH" (default), "MEDIUM", or "LOW"
    """
    gmail_token = await _get_valid_gmail_token(db)
    yahoo_token = await _get_valid_yahoo_token(db)

    if not gmail_token and not yahoo_token:
        raise HTTPException(
            status_code=400,
            detail="No email provider connected. Please connect Gmail or Yahoo Mail first.",
        )

    # Validate confidence param
    if confidence.upper() not in ("HIGH", "MEDIUM", "LOW"):
        confidence = "HIGH"
    confidence = confidence.upper()

    # Load all active jobs for matching
    jobs_result = await db.execute(select(Job).where(Job.is_deleted == False))  # noqa: E712
    jobs = [
        {"id": j.id, "title": j.title, "company": j.company}
        for j in jobs_result.scalars().all()
    ]

    # Load existing message IDs to skip duplicates
    existing_gmail = await db.execute(
        select(ApplicationUpdateSuggestion.gmail_message_id)
    )
    existing_yahoo = await db.execute(
        select(ApplicationUpdateSuggestion.yahoo_message_id)
    )
    existing_gmail_ids = {r for r in existing_gmail.scalars().all() if r}
    existing_yahoo_ids = {r for r in existing_yahoo.scalars().all() if r}

    total_fetched = 0
    total_new = 0
    total_skipped = 0

    # Gmail sync
    if gmail_token:
        try:
            emails = await gmail_fetch(gmail_token.access_token, max_results=50, days_back=7)
            total_fetched += len(emails)
            new, skipped = await _process_emails(
                db, emails, jobs, existing_gmail_ids, "gmail_message_id", confidence
            )
            total_new += new
            total_skipped += skipped
        except Exception:
            pass

    # Yahoo sync
    if yahoo_token:
        try:
            emails = await yahoo_fetch(
                yahoo_token.email, yahoo_token.access_token, max_results=50, days_back=7
            )
            total_fetched += len(emails)
            new, skipped = await _process_emails(
                db, emails, jobs, existing_yahoo_ids, "yahoo_message_id", confidence
            )
            total_new += new
            total_skipped += skipped
        except Exception:
            pass

    await db.commit()

    return {
        "emailsFetched": total_fetched,
        "newSuggestions": total_new,
        "skipped": total_skipped,
        "confidenceLevel": confidence,
    }


@router.get("/suggestions", response_model=list[SuggestionResponse])
async def list_suggestions(
    state: str = "pending",
    db: AsyncSession = Depends(get_db),
):
    """List suggestions filtered by state (default: pending)."""
    query = (
        select(ApplicationUpdateSuggestion)
        .where(ApplicationUpdateSuggestion.state == state)
        .order_by(ApplicationUpdateSuggestion.created_at.desc())
    )
    result = await db.execute(query)
    suggestions = result.scalars().all()

    responses = []
    for s in suggestions:
        job_title = None
        job_company = None
        if s.application_id:
            job = await db.get(Job, s.application_id)
            if job:
                job_title = job.title
                job_company = job.company

        resp = SuggestionResponse.model_validate(s)
        resp.application_title = job_title
        resp.application_company = job_company
        responses.append(resp)

    return responses


@router.patch("/suggestions/{suggestion_id}", response_model=SuggestionResponse)
async def update_suggestion(
    suggestion_id: int,
    payload: SuggestionAction,
    db: AsyncSession = Depends(get_db),
):
    """Apply or ignore a suggestion."""
    suggestion = await db.get(ApplicationUpdateSuggestion, suggestion_id)
    if not suggestion:
        raise HTTPException(status_code=404, detail="Suggestion not found")

    if payload.action == "apply":
        app_id = payload.application_id or suggestion.application_id
        if not app_id:
            raise HTTPException(
                status_code=400,
                detail="No application specified. Choose an application to update.",
            )

        job = await db.get(Job, app_id)
        if not job:
            raise HTTPException(status_code=404, detail="Application not found")
        if job.is_deleted:
            raise HTTPException(status_code=400, detail="Cannot update a deleted application")

        job.status = suggestion.suggested_status
        suggestion.state = "applied"
        suggestion.application_id = app_id

    elif payload.action == "ignore":
        suggestion.state = "ignored"
    else:
        raise HTTPException(status_code=400, detail="Invalid action. Use 'apply' or 'ignore'.")

    await db.commit()
    await db.refresh(suggestion)

    job_title = None
    job_company = None
    if suggestion.application_id:
        job = await db.get(Job, suggestion.application_id)
        if job:
            job_title = job.title
            job_company = job.company

    resp = SuggestionResponse.model_validate(suggestion)
    resp.application_title = job_title
    resp.application_company = job_company
    return resp

File: server/src/routes/emailSync.js
// Email sync routes: run a sync to fetch and process recent Gmail emails,
// list suggestions (pending/applied/ignored), and act on a suggestion.
// All routes require authentication — data is scoped to the logged-in user.

var express = require("express");
var { getPool } = require("../database");
var { formatSuggestionResponse } = require("../helpers");
var gmailService = require("../services/gmailService");
var { processEmail } = require("../services/emailPipeline");
var requireAuth = require("../middleware/auth");

var router = express.Router();

// All email sync routes require a valid login token
router.use(requireAuth);

// Returns the logged-in user's Gmail token, refreshing it if it's about to expire.
// Returns null if the user hasn't connected Gmail.
async function getValidGmailToken(db, userId) {
  var result = await db.query(
    "SELECT * FROM gmail_tokens WHERE user_id = $1 ORDER BY id DESC LIMIT 1",
    [userId]
  );
  // token is the stored Gmail token row, or undefined if not connected
  var token = result.rows[0];
  if (!token) {
    return null;
  }

  // Check if the token expires within the next 5 minutes
  var fiveMinutesFromNow = new Date(Date.now() + 5 * 60 * 1000);
  if (token.token_expiry && new Date(token.token_expiry) < fiveMinutesFromNow) {
    try {
      // refreshed contains the new access_token and optionally a new refresh_token
      var refreshed = await gmailService.refreshAccessToken(token.refresh_token);
      var newExpiry = new Date(Date.now() + (refreshed.expires_in || 3600) * 1000);

      // newRefreshToken is the updated refresh token (sometimes Google rotates them)
      var newRefreshToken = refreshed.refresh_token || token.refresh_token;

      await db.query(
        `UPDATE gmail_tokens
         SET access_token = $1, token_expiry = $2, refresh_token = $3, updated_at = NOW()
         WHERE id = $4`,
        [refreshed.access_token, newExpiry, newRefreshToken, token.id]
      );

      token.access_token = refreshed.access_token;
      token.token_expiry = newExpiry;
    } catch (e) {
      return null;
    }
  }

  return token;
}

// Runs fetched emails through the pipeline and inserts new suggestions for this user.
// Returns an object { newSuggestions, skipped } with counts.
async function processEmails(db, userId, emails, jobs, existingIds, minConfidence) {
  var newSuggestions = 0;
  var skipped = 0;

  for (var i = 0; i < emails.length; i++) {
    var emailData = emails[i];
    // msgId is the unique identifier for this email from Gmail
    var msgId = emailData.id;

    // Skip emails we've already processed
    if (existingIds.has(msgId)) {
      skipped++;
      continue;
    }

    // Run the email through the full pipeline (pre-filter, score, match, classify)
    var pipelineResult = processEmail(emailData, jobs, minConfidence);
    if (pipelineResult === null) {
      continue;
    }

    await db.query(
      `INSERT INTO application_update_suggestions
        (user_id, application_id, detected_type, suggested_status, confidence_score,
         email_subject, email_sender, email_date, email_snippet, gmail_message_id, state)
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, 'pending')`,
      [
        userId,
        pipelineResult.matchedApplicationId,
        pipelineResult.classification,
        pipelineResult.suggestedStatus,
        pipelineResult.confidence,
        pipelineResult.emailSubject,
        pipelineResult.emailSender,
        pipelineResult.emailDate,
        pipelineResult.emailSnippet,
        msgId,
      ]
    );
    newSuggestions++;
  }

  return { newSuggestions: newSuggestions, skipped: skipped };
}

// Fetch recent emails from Gmail, run them through the detection pipeline,
// and create suggestions for the logged-in user
router.post("/run", async function (req, res, next) {
  try {
    var db = getPool();
    var userId = req.user.id;

    // confidence is the minimum confidence level from the query string
    var confidence = (req.query.confidence || "HIGH").toUpperCase();
    if (confidence !== "HIGH" && confidence !== "MEDIUM" && confidence !== "LOW") {
      confidence = "HIGH";
    }

    var gmailToken = await getValidGmailToken(db, userId);

    if (!gmailToken) {
      res.status(400).json({
        detail: "Gmail is not connected. Please connect Gmail first.",
      });
      return;
    }

    // Load this user's active (non-deleted) jobs for matching
    var jobsResult = await db.query(
      "SELECT id, title, company FROM jobs WHERE user_id = $1 AND is_deleted = false",
      [userId]
    );
    var jobs = jobsResult.rows;

    // Load existing Gmail message IDs for this user so we can skip duplicates
    var existingResult = await db.query(
      "SELECT gmail_message_id FROM application_update_suggestions WHERE user_id = $1 AND gmail_message_id IS NOT NULL",
      [userId]
    );

    // existingIds is a Set of Gmail message IDs we've already processed for this user
    var existingIds = new Set(
      existingResult.rows.map(function (r) { return r.gmail_message_id; })
    );

    var totalFetched = 0;
    var totalNew = 0;
    var totalSkipped = 0;

    // Fetch and process Gmail emails
    try {
      var emails = await gmailService.fetchRecentEmails(gmailToken.access_token, 50, 7);
      totalFetched = emails.length;

      var results = await processEmails(db, userId, emails, jobs, existingIds, confidence);
      totalNew = results.newSuggestions;
      totalSkipped = results.skipped;
    } catch (e) {
      // silently continue if Gmail fetch fails
    }

    res.json({
      emailsFetched: totalFetched,
      newSuggestions: totalNew,
      skipped: totalSkipped,
      confidenceLevel: confidence,
    });
  } catch (err) {
    next(err);
  }
});

// List suggestions for the logged-in user, filtered by state (default: "pending")
router.get("/suggestions", async function (req, res, next) {
  try {
    var db = getPool();
    var userId = req.user.id;
    // state is the filter — "pending", "applied", or "ignored"
    var state = req.query.state || "pending";

    var result = await db.query(
      "SELECT * FROM application_update_suggestions WHERE user_id = $1 AND state = $2 ORDER BY created_at DESC",
      [userId, state]
    );

    var suggestions = result.rows;
    var responses = [];

    // For each suggestion, look up the matched job to include its title and company
    for (var i = 0; i < suggestions.length; i++) {
      var s = suggestions[i];
      var jobTitle = null;
      var jobCompany = null;

      if (s.application_id) {
        var jobResult = await db.query(
          "SELECT title, company FROM jobs WHERE id = $1 AND user_id = $2",
          [s.application_id, userId]
        );
        // matchedJob is the job record this suggestion refers to, if it still exists
        var matchedJob = jobResult.rows[0];
        if (matchedJob) {
          jobTitle = matchedJob.title;
          jobCompany = matchedJob.company;
        }
      }

      responses.push(formatSuggestionResponse(s, jobTitle, jobCompany));
    }

    res.json(responses);
  } catch (err) {
    next(err);
  }
});

// Apply or ignore a suggestion for the logged-in user.
// If "apply", it updates the matched job's status to the suggested one.
// If "ignore", it just marks the suggestion as ignored.
router.patch("/suggestions/:suggestionId", async function (req, res, next) {
  try {
    var db = getPool();
    var userId = req.user.id;
    // suggestionId is the ID from the URL path
    var suggestionId = parseInt(req.params.suggestionId, 10);

    var result = await db.query(
      "SELECT * FROM application_update_suggestions WHERE id = $1 AND user_id = $2",
      [suggestionId, userId]
    );
    var suggestion = result.rows[0];

    if (!suggestion) {
      res.status(404).json({ detail: "Suggestion not found" });
      return;
    }

    // action is either "apply" or "ignore", sent in the request body
    var action = req.body.action;
    // applicationId can be sent to override which job to update
    var applicationId = req.body.applicationId || req.body.application_id;

    if (action === "apply") {
      // appId is the ID of the job to update — either from the request or from the suggestion
      var appId = applicationId || suggestion.application_id;

      if (!appId) {
        res.status(400).json({
          detail: "No application specified. Choose an application to update.",
        });
        return;
      }

      var jobResult = await db.query(
        "SELECT * FROM jobs WHERE id = $1 AND user_id = $2",
        [appId, userId]
      );
      var job = jobResult.rows[0];

      if (!job) {
        res.status(404).json({ detail: "Application not found" });
        return;
      }
      if (job.is_deleted) {
        res.status(400).json({ detail: "Cannot update a deleted application" });
        return;
      }

      // Update the job's status to what the suggestion recommends
      await db.query(
        "UPDATE jobs SET status = $1, updated_at = NOW() WHERE id = $2 AND user_id = $3",
        [suggestion.suggested_status, appId, userId]
      );

      // Mark the suggestion as applied and link it to the job
      await db.query(
        "UPDATE application_update_suggestions SET state = 'applied', application_id = $1 WHERE id = $2 AND user_id = $3",
        [appId, suggestionId, userId]
      );

      suggestion.state = "applied";
      suggestion.application_id = appId;

    } else if (action === "ignore") {
      await db.query(
        "UPDATE application_update_suggestions SET state = 'ignored' WHERE id = $1 AND user_id = $2",
        [suggestionId, userId]
      );
      suggestion.state = "ignored";

    } else {
      res.status(400).json({ detail: "Invalid action. Use 'apply' or 'ignore'." });
      return;
    }

    // Look up the matched job to include in the response
    var jobTitle = null;
    var jobCompany = null;

    if (suggestion.application_id) {
      var jobLookup = await db.query(
        "SELECT title, company FROM jobs WHERE id = $1 AND user_id = $2",
        [suggestion.application_id, userId]
      );
      var matchedJob = jobLookup.rows[0];
      if (matchedJob) {
        jobTitle = matchedJob.title;
        jobCompany = matchedJob.company;
      }
    }

    res.json(formatSuggestionResponse(suggestion, jobTitle, jobCompany));
  } catch (err) {
    next(err);
  }
});

module.exports = router;
